src/parser/lexer.rs

```diff
@@ -0,0 +1,140 @@
+use std::{iter::Peekable, str::Chars};
+
+use super::token::Token;
+use super::range::{Range, Position};
+
+/// The lexer state
+#[derive(Debug)]
+pub struct Lexer<'a> {
+    pub input: &'a str,
+    pub position: usize,
+    pub peekable: &'a mut Peekable<Chars<'a>>
+}
+
+fn is_whitespace(c: char) -> bool {
+    matches!(c, ' ' | '\r' | '\t')
+}
+
+impl<'a> Lexer<'a> {
+    pub fn new(
+        input: &'a str,
+        peekable: &'a mut Peekable<Chars<'a>>
+    ) -> Lexer {
+        Lexer {
+            input,
+            position: 0,
+            peekable,
+        }
+    }
+
+    pub fn tokenize(&mut self) -> (Token, Range) {
+        let start = self.position;
+
+        match self.peekable.peek() {
+            None => (Token::EOF, self.mk_range(start)),
+            Some(c) => match c {
+                c if is_whitespace(*c) => {
+                    self.accumulate_while(&is_whitespace);
+                    self.tokenize()
+                }
+                // TODO: Add other types of numbers on it too
+                c if c.is_ascii_digit() => {
+                    let num = self.accumulate_while(&|x| x.is_ascii_digit() || x == '_');
+                    (Token::LInt(num.replace('_', "").parse().unwrap()), self.mk_range(start))
+                }
+                '(' => self.to_single_token(Token::LPar, start),
+                ')' => self.to_single_token(Token::RPar, start),
+                '+' => self.to_single_token(Token::OpPlus, start),
+                _ => unreachable!(),
+            }
+        }
+    }
+
+    fn to_single_token(&mut self, token: Token, start: usize) -> (Token, Range) {
+        self.next_char();
+        (token, self.mk_range(start))
+    }
+
+    fn next_char(&mut self) -> Option<char> {
+        match self.peekable.next() {
+            Some(c) if !self.input.is_empty() => {
+                self.input = &self.input[c.len_utf8()..];
+                self.position += c.len_utf8();
+                Some(c)
+            }
+            _ => None
+        }
+    }
+
+    fn mk_range(&self, start: usize) -> Range {
+        Range::new(
+            Position { index: start as u32 },
+            Position { index: self.position as u32 }
+        )
+    }
+
+    fn accumulate_while(&mut self, condition: &dyn Fn(char) -> bool) -> &str {
```

Comment:
Just out of curiosity's sake, what made you use `dyn` here?

Reply:
It's another thing that comes from the kind-parser lol. I had to use it in the original code base in order to make some changes on the current scope, thus requiring a closure.

Reply:
Why not `impl`?

Reply:
How it could be done with `impl`?